File: 0133_Clone Graph/CloneGraph_DFS.cs
public class Solution {
    private Dictionary<Node, Node> graph = new Dictionary<Node, Node>();
    
    public Node CloneGraph(Node node) {
        if(node == null) return null;
        if(graph.ContainsKey(node)) return graph[node];
        
        var newNode = new Node(node.val);
        graph.Add(node, newNode);
        
        foreach(var nei in node.neighbors)
        {
            newNode.neighbors.Add(CloneGraph(nei));
        }
        
        return newNode;
    }
    
    
}

File: 0133_Clone Graph/CloneGraph_DFS.java
class Solution {
    public Node cloneGraph(Node node) {
        if(node == null) return null;
        Map<Node, Node> map = new HashMap<>();
        DFS(node, map);
        return map.get(node);
    }
    
    private void DFS(Node node, Map<Node, Node> map) {
        if(node == null || map.containsKey(node)) return;
        Node copy = new Node(node.val);
        map.put(node, copy);
        for(Node next : node.neighbors) {
            DFS(next, map);
            copy.neighbors.add(map.get(next));
        }
    }
}